pyfr/plugins/base.py

```diff
@@ -167,3 +169,90 @@ def __init__(self, intg, *args, **kwargs):
 
             if not isinstance(eidxs, slice):
                 self._ele_region_data[f'{etype}_idxs'] = eidxs
+
+
+class ElementMixin:
+    def __init__(self, intg, *args, **kwargs):
+        super().__init__(intg, *args, **kwargs)
+
+        self._surf_quad = {}
+
+    def surf_quad(self, itype, proj, *args):
+        try:
+            ppts, pwts = self._surf_quad[itype, proj]
+        except KeyError:
+            # Obtain quadrature info
+            rname = self.cfg.get(f'solver-interfaces-{itype}', 'flux-pts')
+
+            # Quadrature rule (default to that of the solution points)
+            qrule = self.cfg.get(self.cfgsect, f'quad-pts-{itype}', rname)
+            try:
+                qdeg = self.cfg.getint(self.cfgsect, f'quad-deg-{itype}')
+            except NoOptionError:
+                qdeg = self.cfg.getint(self.cfgsect, 'quad-deg')
+
+            q = get_quadrule(itype, qrule, qdeg=qdeg, flags='s')
+            ppts, pwts = self._proj_pts(proj, q.pts), q.wts
+            self._surf_quad[itype, proj] = (ppts, pwts)
+
+        return ppts, pwts
+
+    @staticmethod
+    def _proj_pts(projector, pts):
+        pts = np.atleast_2d(pts.T)
+        return np.vstack(np.broadcast_arrays(*projector(*pts))).T
+
+    @staticmethod
+    def plocs_at(eles, eidx, pts):
+        nspts, neles, ndims = eles.nspts, len(eidx), eles.ndims
+
+        # Construct the physical location operator matrix
+        plocop = eles.basis.sbasis.nodal_basis_at(pts)
+
+        # Apply the operator to the mesh elements and reshape
+        e_surf = eles.eles[:, eidx]
+        ploc = plocop @ e_surf.reshape(nspts, -1)
+        ploc = ploc.reshape(len(pts), neles, ndims)
+
+        return ploc
+
+    @staticmethod
+    def jacs_at(eles, eidx, pts):
```

Comment:
There looks to be a lot of code duplication here.  It should be possible to eliminate most of this by extending some of the eles `_at_np` methods to support more flexible `name` arguments (see `rcpdjac_at_np` for an example).